### DISCUSSION TITLE
App doesn't reload on new deployment

### DISCUSSION THREAD

**Discussion by pratv1** [vercel/next.js#59099] [Help]
### Summary

In the [docs](https://nextjs.org/docs/pages/building-your-application/deploying#version-skew) it says:
> Next.js will automatically mitigate most instances of version skew and automatically reload the application to retrieve new assets when detected. For example, if there is a mismatch in the build ID, transitions between pages will perform a hard navigation versus using a prefetched value.

But this is not working in my case. The app doesn't reload when I navigate to a different page after making a new deployment. Also, the `getServerSideProps` function returns `undefined`. To fix it, I have to manually refresh the page.

I'm using the pages router btw.

### Additional information

```js
next version: 14.0.3
node version: 20.9.0
```

**Comment by Rathalin** [comment]
I'm experiencing the same problem. There seems to be an issue with a custom [middleware ](https://nextjs.org/docs/app/building-your-application/routing/middleware) because the app correctly reloads without a middleware file.

**Comment by pratv1** [comment]
Nope, I'm not using any middleware.

Anyway, if it helps, to fix this I ended up manually generating a build ID, then comparing the client and server build IDs with a `useEffect` inside the root component. This will make it reload the app when a user navigates to a different page and the build IDs don't match.

```js
// next.config.js

const buildId = new Date().getTime().toString();
const nextConfig = {
  env: {
    NEXT_PUBLIC_BUILD_ID: buildId,
  },
  // ...
}
```

```js
// pages/api/build-id.js

export default async function handler(req, res) {
  if (req.method === "GET") {
    res.status(200).json({ buildId: process.env.NEXT_PUBLIC_BUILD_ID });
  } else {
    res.status(405).json({ success: false, error: "Method not allowed" });
  }
}
```

```js
// pages/_app.js

function App() {
  // Force reload if build IDs don't match
  useEffect(() => {
    fetch("/api/build-id")
      .then((res) => res.json())
      .then((data) => {
        const clientBuildId = process.env.NEXT_PUBLIC_BUILD_ID;
        const serverBuildId = data.buildId;
        if (clientBuildId !== serverBuildId) {
          window?.location.reload();
        }
      });
  });
  // ...
}
```

**Comment by Rathalin** [comment]
Thank you! I'll probably use your workaround aswell.

**Comment by Rathalin** [comment]
I found a simple workaround:

```ts
// middleware.page.ts
export function middleware(req: NextRequest) {
  if (req.url.includes('_next/data')) {
    return new NextResponse(null, { status: 404 });
  }
  ...
}
```

From my understanding, requests that include `_next/data/<BUILD_ID>/` in the url should never reach this middleware. However, it seems that only requests with the current `BUILD_ID` are filtered but requests containing an old `BUILD_ID` somehow end up in this middleware.
Normally, Next.js automatically sends a `404` response, if the `BUILD_ID` doesn't match. Here, I manually send a `404`, which means the browser/app will perform a full page reload and users always have the latest version of the app.

Unfortunately for this case, the [docs ](https://nextjs.org/docs/pages/building-your-application/deploying#version-skew) have been updated and no longer explain this behaviour.

**Reply by Rathalin** [reply]
To be clear, I think this is a bug in Next.js.

**Reply by Mitchnsun** [reply]
Thanks for this workaround, it works like a charm!
I got this issue for time on each deployment. I tried to generate a build id (https://nextjs.org/docs/app/building-your-application/deploying#build-cache) but does not change anything to the issue

**Reply by Afnan-Riaz** [reply]
@Rathalin I am using next 13. The workaround you have suggested may solve the issue. but when i search on the internet, people are complaining about the 404 responses for _next/data requests. Wouldn't your solution generate more 404 errors on the console?

**Reply by Rathalin** [reply]
I don't understand how a 404 response would be a problem on the console or in general.

**Reply by sayoojtmc** [reply]
@Rathalin Can we use the matcher regex exported from the middleware to ignore paths that include `_next/data` ?

**Reply by sayoojtmc** [reply]
I've tried adding `_next/data` to the matcher array in middleware, but when we rebuild the page, the `_next/data` is somehow getting into the middleware. It does not enter the middle after that though.

**Reply by Rathalin** [reply]
In my experience, adding `_next/data` to the matcher does not help.
